nucypher/blockchain/economics.py

```diff
@@ -14,60 +14,167 @@
+    hours_per_period = 24                            # Hours in single period
+    seconds_per_period = hours_per_period * 60 * 60  # Seconds in single period
+
+    # Time Constraints
+    minimum_locked_periods = 30  # 720 Hours minimum
 
-class Economics:
-    def __init__(
-            self,
-            initial_supply=10**9,
-            initial_inflation=1.0,
-            T_half=2.0,
-            T_sat=1.0,
-            small_staker_multiplier=0.5) -> None:
+    # Value Constraints
+    minimum_allowed_locked = NU(15_000, 'NU').to_nunits()
+    maximum_allowed_locked = NU(4_000_000, 'NU').to_nunits()
+
+    # Supply
+    __default_initial_supply = NU(int(1_000_000_000), 'NU').to_nunits()
+    __default_initial_inflation = 1
+    __default_token_halving = 2      # years
+    __default_reward_saturation = 1  # years
+    __default_small_stake_multiplier = Decimal(0.5)
+
+    def __init__(self,
+                 initial_supply: int = __default_initial_supply,
+                 initial_inflation: int = __default_initial_inflation,
+                 halving_delay: int = __default_token_halving,
+                 reward_saturation: int = __default_reward_saturation,
+                 small_stake_multiplier: Decimal = __default_small_stake_multiplier
+                 ):
         """
-        Calculate parameters to use in token issuer from high-level
-        human-understandable parameters
-
-        :param initial_supply: coins at t=0,
-        :param initial_inflation: inflation in day 1 expressed in units of year**-1
-            e.g. 1/365 of all coins in day 1 is 1 year**-1
-        :param T_half: time for inflation halving in years
-        :param T_sat: "saturation" time - if staking is longer than T_sat, the
-            reward doesn't get higher
-        :param small_staker_multiplier: what fraction of max reward rate do we
-            pay to those who are about to unlock coins
+        :param initial_supply: Tokens at t=0
+        :param initial_inflation: Inflation on day 1 expressed in units of year**-1
+        :param halving_delay: Time for inflation halving in years
+        :param reward_saturation: "saturation" time - if staking is longer than T_sat, the reward doesn't get any higher
+        :param small_stake_multiplier: Fraction of maximum reward rate paid to those who are about to unlock tokens
         """
 
-        # Saving input in case we'll need it
-        self.input_params = dict(
-                initial_supply=initial_supply,
-                initial_inflation=initial_inflation,
-                T_half=T_half,
-                T_sat=T_sat,
-                small_staker_multiplier=small_staker_multiplier)
+        #
+        # Calculate
+        #
+
+        with localcontext() as ctx:
+            ctx.prec = self._precision
+
+            initial_supply = Decimal(initial_supply)
+
+            # ERC20 Token parameter
+            total_supply = initial_supply * (1 + initial_inflation * halving_delay / LOG2)
```

Comment:
We can add as a comment here that this refers to Equation 4 in the mining paper:
```
# ERC20 Token parameter (See Equation 4 in Mining paper)
```

Reply:
:heavy_check_mark: